### DISCUSSION TITLE
Adding more options in builtin DNS providers

### DISCUSSION THREAD

**Discussion by agpt8** [HotCakeX/WinSecureDNSMgr#5] [Ideas]
Hi there,

From what I can tell from the README, currently, Cloudflare, Quad9 and Google are supported in builtin DNS provider list. Specifically, the default values for cloudflare and quad9 (I mean, not the malware blocking ones). Would you consider adding those under the builtin provider list?

For cloudflare, the DNS list available [here](https://developers.cloudflare.com/1.1.1.1/setup/windows/).
For Quad9, it can be found [here](https://quad9.net/service/service-addresses-and-features).

I know these can be added manually with the custom dns cmdlet in the repo but due to lack of docs, it is a bit tricky and requires a bit of fiddling around. For people not familiar with powershell, it would be tricky to find all the acceptable parameter values. For me, using the Get-Help cmdlet did the trick `Get-Help Set-CustomWinSecureDNS -Full`.

Also, would you consider the fallback to be reordered to cloudflare > quad9 > google for the failed queries? This is in reference to the failed IP lookup on DoH servers as mentioned under [features](https://github.com/HotCakeX/WinSecureDNSMgr?tab=readme-ov-file#features). I ask this as cloudflare and quad9 are far more privacy centric than google even though the all of the queries are encrypted  (think no/minimal logs as compared to google). The best case scenario would be having an option in the script itself so that users themselves can mention the order they want.

**Comment by HotCakeX** [comment]
Hi,
those are good suggestions, i'll implement all of them!

**Comment by HotCakeX** [comment]
I added Cloudflare family, Cloudflare malware and Quad9 Secured w/ECS: Malware blocking to the list of the supported providers for the built-in DoHs.

I couldn't find the direct IP address resolver for Quad9 to use that as a fallback instead of Google. Direct IP addresses are needed because when the fallbacks are used it's assumed that the system DNS is failing so domain names cannot be resolved at that point. This shouldn't matter much though because Cloudflare's 2 main IP addresses are always tried first so Google is only used if Cloudflare IPs are blocked in your network which should be really rare.
